node-graph/libraries/vector-types/src/vector/misc.rs

```diff
@@ -246,6 +246,74 @@ pub fn pathseg_abs_diff_eq(seg1: PathSeg, seg2: PathSeg, max_abs_diff: f64) -> b
 
 	seg1_points.len() == seg2_points.len() && seg1_points.into_iter().zip(seg2_points).all(|(a, b)| cmp(a.x, b.x) && cmp(a.y, b.y))
 }
+pub trait Tangent {
+	fn tangent_at(&self, t: f64) -> DVec2;
+
+	fn tangent_at_start(&self) -> DVec2 {
+		self.tangent_at(0.0)
+	}
+
+	fn tangent_at_end(&self) -> DVec2 {
+		self.tangent_at(1.0)
+	}
+}
+
+impl Tangent for kurbo::Line {
+	fn tangent_at(&self, _t: f64) -> DVec2 {
+		self.tangent_at_start()
+	}
+
+	fn tangent_at_start(&self) -> DVec2 {
+		let delta = self.p1 - self.p0;
+		DVec2 { x: delta.x, y: delta.y }
+	}
+
+	fn tangent_at_end(&self) -> DVec2 {
+		self.tangent_at_start()
+	}
+}
+
+impl Tangent for kurbo::QuadBez {
+	fn tangent_at(&self, t: f64) -> DVec2 {
+		let deriv = self.deriv();
+		let tangent = deriv.eval(t);
+		DVec2 { x: tangent.x, y: tangent.y }
```

Comment:
If the handle is placed on top of the anchor then this will be zero.

For reference, here is my impl of the tangent at start which attempts to avoid reporting zero length tangents:
https://github.com/GraphiteEditor/Graphite/blob/06749ecea52f3085903bca5a0b242f3a83f9538c/node-graph/libraries/vector-types/src/vector/algorithms/bezpath_algorithms.rs#L704-L724